tests/test_client.py

```diff
@@ -163,46 +193,59 @@ def test_client_load_cities_as_dictionary(city_client_and_get_mock):
     assert cities[0]["state"]["name"] == "Estado da Guanabara"
 
 
-def test_client_load_cities_with_city_id(city_client_and_get_mock):
+@mark.asyncio
+async def test_async_client_load_cities_with_city_id(city_client_and_get_mock):
     client, mock = city_client_and_get_mock
-    client.cities(city_id="21")
+    await client.cities(city_id="21")
     mock.assert_called_once_with(
         "http://127.0.0.1/api/v2/cities?cityId=21",
         headers={"Authorization": "Bearer 42"},
     )
 
 
-def testclientes_with_city_name(city_client_and_get_mock):
+@mark.asyncio
+async def test_async_client_with_city_name(city_client_and_get_mock):
     client, mock = city_client_and_get_mock
-    client.cities(city_name="Rio de Janeiro")
+    await client.cities(city_name="Rio de Janeiro")
     mock.assert_called_once_with(
         "http://127.0.0.1/api/v2/cities?cityName=Rio+de+Janeiro",
         headers={"Authorization": "Bearer 42"},
     )
 
 
-def test_client_load_cities_with_state_id(city_client_and_get_mock):
+@mark.asyncio
+async def test_async_client_load_cities_with_state_id(city_client_and_get_mock):
     client, mock = city_client_and_get_mock
-    client.cities(state_id="42")
+    await client.cities(state_id="42")
     mock.assert_called_once_with(
         "http://127.0.0.1/api/v2/cities?stateId=42",
         headers={"Authorization": "Bearer 42"},
     )
 
 
-def test_client_load_cities_with_more_than_one_params(city_client_and_get_mock):
+@mark.asyncio
+async def test_async_client_load_cities_with_more_than_one_params(
+    city_client_and_get_mock,
+):
     client, mock = city_client_and_get_mock
-    client.cities(state_id="42", city_name="Rio de Janeiro")
+    await client.cities(state_id="42", city_name="Rio de Janeiro")
     mock.assert_called_once_with(
         "http://127.0.0.1/api/v2/cities?cityName=Rio+de+Janeiro&stateId=42",
         headers={"Authorization": "Bearer 42"},
     )
 
 
-def test_client_occurrences(occurrences_client_and_get_mock):
+@mark.asyncio
+async def test_async_client_occurrences(occurrences_client_and_get_mock):
     client, mock = occurrences_client_and_get_mock
-    client.occurrences(42)
+    await client.occurrences(42)
     mock.assert_called_once_with(
         "http://127.0.0.1/api/v2/occurrences?idState=42&typeOccurrence=all&page=1",
         headers={"Authorization": "Bearer 42"},
     )
+
+
+@patch.object(Client, "states")
+def test_client_load_states(states_mock):
+    Client.states(format=None)
+    states_mock.assert_called_with(format=None)
```

Comment:
Esse teste está testando o _mock_, não o código que você escreveu. Ele é igual a:

```python
mock = Mock()
mock(42)
assert.mock_called_once_with(42)
```

Reply:
Fiz alguns testes, mas não cheguei a nenhum lugar.
Entendo que, se devemos mockar o AsynClient para testar se o `Client` esá usando o seu método `AsyncClient.states`, poderíamos usar o que já temos construido, ex: `state_client_and_get_mock`, que nos retorna um `client` (instancia de `AsynClient` mockado) e um resultado esperado.
imagino que seria algo como:
```
def test_client_load_states(state_client_and_get_mock):
    async_client, mock = state_client_and_get_mock 
    Client().states(format=None)
    asser async_client.states.called
```
Mas essa tentativa falha com um `StopIteration`.

Revisando o mock mencionado, vejo que o `state_client_and_get_mock` está mockando o método `get` de uma instancia `AsyncClient`, fazendo ele retornar o valor esperado.
Então estou chegando a con~fusão~clusão de queria o caso de mockar o método `states` de `AsynClient` para usar-lo e confirmar se está sendo usado. 
Faz sentido isso?
Algo como:
```
# conftest
@fixture
def state_client_mock(client_with_token):
    data = {"data": [{"id": "42", "name": "Rio de Janeiro"}]}
    with async_respond_with("states", data) as states:
        yield (client_with_token, states)
```
Acho que estou perdido....

Reply:
> Mas essa tentativa falha com um `StopIteration`.

Negativo. Esse código não gera um `NameError: name 'asser' is not defined`. Pode me passar o link do commit com o código que gera essa erro? [Deixando esse lembrete aqui](https://github.com/FelipeSBarros/crossfire/pull/74#issuecomment-1803111560) hehehe… 😈 

---

No mais, a minha dica por agora seria: esqueça _fixture_ e o `conftest.py`. Faça algo simples. Depois, quando estiver funcionando, a gente generaliza. Se a gente não tivesse o `conftest.py`, como implementaria o que eu [sugeri](https://github.com/FelipeSBarros/crossfire/pull/74#issuecomment-1818990740):

> você pode só _mockar_ a classe `AsyncClient` e esperar que ela seja chamada com os argumentos desejados. 

---

> Acho que estou perdido....

Você complicou demais. Lembre [do que eu disse antes](https://github.com/FelipeSBarros/crossfire/pull/74#issuecomment-1818990740):


> Não vejo necessidade de uma nova _fixture_.

Imagine que eu tenha uma função:

```python
# app.py

def magic():
    client = AsyncClient()
    return client.answer(42)
```

Como eu faço um teste que?
* faz um _mock_ da classe `AsyncClient`
* verifica que ela foi chamada com com `42`

Provavelmente algo simples com um `with patch("app.AsyncClient") as client`, não?

Reply:
@cuducos revendo a sua [live sobre mock](), logo no início você apresenta alguns exemplos como instroução.
Em base a esse exercício proposto imagino que seja algo como:
```Python
AsyncClient = Mock()
AsyncClient.answer.return_value = 42

result = magic()

# verificação
AsyncClient.assert_called_once()
```
Mas o `AsyncClient.assert_called_once()` retorna `None`. Logo, algo não está implementado como deveria.

Contudo, se eu uso `AsyncClient.called_count()`, ma indica que foi chamado uma vez.
Ao executar `magic()` mais uma vez o `AsyncClien.called_count` é incrementado.

Reply:
Lendo a documentação sobre [mock de classes](https://docs.python.org/3/library/unittest.mock-examples.html#mocking-classes), vi um exemplo, ao qual tentei adaptar ao caso do que estamos tentando testar.
O que está na doc:
```python
def some_function():
    instance = module.Foo()
    return instance.method()

with patch('module.Foo') as mock:
    instance = mock.return_value
    instance.method.return_value = 'the result'
    result = some_function()
    assert result == 'the result'
```

Minha adaptação:
```python
def test_client_load_states():
    with patch("crossfire.client.AsyncClient") as mock:
        async_client = mock
        async_client.sates.return_value = 42
        client_ = Client()
        client_.states()
        async_client.states.assert_called()
```
Isso gera a seguinte falha:
```python
FAILED tests/test_client.py::test_client_load_states - AttributeError: <functools._lru_cache_wrapper object at 0x7f527c43fe20> does not have the attribute 'AsyncClient'
```

Reply:
Fiz uma pequena alteração pois acho que a adaptação que fiz estava tnetando acessar a classe `AsyncClient` em `crossfire.client` e não `crossfire.clients`, como deveria:
```python

def test_client_load_states():
    with patch("crossfire.clients.AsyncClient") as mock:
        async_client = mock
        async_client.sates.return_value = 42
        client = Client()
        client.states()
        async_client.states.assert_called()

```

Reply:
Beleza. Assim que você comittar o código, eu olho, como disse [aqui](https://github.com/FelipeSBarros/crossfire/pull/74#issuecomment-1803111560) ; )

Reply:
Uma estratégia que te ajuda por esse caminho é o [`patch.object`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch.object).

Reply:
> Beleza. Assim que você comittar o código, eu olho, como disse [aqui](https://github.com/FelipeSBarros/crossfire/pull/74#issuecomment-1803111560) ; )

Blz. Não havia commitado antes pois estava perdido e tentando, antes, ter uma ideia da direção a seguir.
Commit feito.

Reply:
> Não havia commitado antes pois estava perdido e tentando, antes, ter uma ideia da direção a seguir.

O problema é que por [duas](https://github.com/FelipeSBarros/crossfire/pull/74#discussion_r1403640487) [vezes](https://github.com/FelipeSBarros/crossfire/pull/74#discussion_r1405246405) você compartilhava código como comentário que, quando executado aqui (sim, eu copiei e colei), gerava algo diferente to comentário em si.

Se não consigo replicar o que vc está relatando, me sentia perdido em _como_ ajudar. Normalmente os erros em si nos mostram o caminho, seja erro de sintaxe ou do teste em si. Se a gente está tendo erros diferentes com o mesmo código, fica difícil alinhar o próximo passo.

Reply:
Complementando: se a gente tem o mesmo erro, a gente pode quebrar o problema maior em passos pequenos, focar em um erro por vez, e ir progredingo assim. O alinhamento que precisávamos, acho que já foi feito, você mesmo resumiu aqui:

> Entendo que, se devemos mockar o `AsynClient` para testar se o `Client` esá usando o seu método `AsyncClient.states`